### PR TITLE
Added deepspeed to the latest megatron-lm

### PR DESCRIPTION
**Pull request by cli99**

### CODE REVIEW

File: megatron/arguments.py [hunk 1]
```diff
@@ -270,55 +347,79 @@ def _add_initialization_args(parser):
...
+        'from checkpoint and ignore input arguments.')
 
     return parser
 
 
 def _add_checkpointing_args(parser):
     group = parser.add_argument_group(title='checkpointing')
 
-    group.add_argument('--save', type=str, default=None,
+    group.add_argument('--save',
```
Comment (ShadenSmith) [inline megatron/arguments.py]: It looks like auto-formatting happened at some point and it's blowing up the diff size :-(. I wonder if we can quickly revert these by running whatever formatter Megatron uses?

**Comment by msmolyak** [comment]
I tried running model pre-training with this branch in a Docker container used by NVIDIA for Megatron-LM (nvcr.io/nvidia/pytorch:20.03-py3) with the latest dev version of DeepSpeed installed and was able to successfully perform the operation (on a single GPU machine).

One aspect of Megatron-LM not updated with DeepSpeed wrapper is text generation after model training. It would be helpful for the Microsoft team to modify [generate_samples_gpt2.py](https://github.com/cli99/DSE/blob/cheng/megatron-deepspeed/tools/generate_samples_gpt2.py) program to wrap the original Megatron-LM model with DeepSpeed engine. As is, this program fails since it cannot load the checkpoint generated by DeepSpeed using Megatron-LM model.

Here is the modification, which allowed me to use the text generation program against the DeepSpeed-generated model:

```
def main():
    """Main program."""

    initialize_megatron(extra_args_provider=add_text_generate_args,
                        args_defaults={'tokenizer_type': 'GPT2BPETokenizer'})

    # Set up model and load checkpoint.
    model = get_model(model_provider)
    args = get_args()

    if args.deepspeed and args.fp16:
        model.half()

    if args.deepspeed:
        print_rank_0("DeepSpeed is enabled.")

        model = deepspeed.initialize(
            model=model,
            optimizer=None,
            args=args,
            lr_scheduler=None,
            mpu=None,
            dist_init_required=False
        )[0]

    if args.load is not None:
        _ = load_checkpoint(model, None, None)

```

One caveat is that I had to make a change to DeepSpeed code in order for the deepspeed.initialize() call above to work without an optimizer. The change is this: [https://github.com/microsoft/DeepSpeed/pull/469/files](https://github.com/microsoft/DeepSpeed/pull/469/files)

My team is trying to use the latest Megatron-LM code with DeepSpeed and it would help us greatly if you approved this no-optimizer pull request and added the change to generate_samples_gpt2.py to this pull request (or added it separately shortly after this one is merged).

File: megatron/checkpointing.py [hunk 2]
```diff
@@ -89,45 +89,68 @@ def get_checkpoint_tracker_filename(checkpoints_path):
...
-        checkpoint_name = get_checkpoint_name(args.save, iteration)
-        print('global rank {} is saving checkpoint at iteration {:7d} to {}'.
-              format(torch.distributed.get_rank(), iteration, checkpoint_name))
-        ensure_directory_exists(checkpoint_name)
-        torch.save(state_dict, checkpoint_name)
-        print('  successfully saved {}'.format(checkpoint_name))
+    #megatron model uses state_dict_for_save_checkpointing instead of the standard state_dict
+    #state_dict is used by deepspeed for module saving so it needs to point to the right function
+    model.module.state_dict = model.module.state_dict_for_save_checkpoint
+    model.save_checkpoint(args.save, iteration, client_state=sd)
```
Comment (gcooper-isi) [inline megatron/checkpointing.py]: I get the following exception when running `pretrain_gpt2.py` on this branch with DeepSpeed version 0.3.10:

```
Traceback (most recent call last):
  File "pretrain_gpt2.py", line 115, in <module>
    args_defaults={'tokenizer_type': 'GPT2BPETokenizer'})
  File "/data/DSE/megatron/training.py", line 101, in pretrain
    train_data_iterator, valid_data_iterator)
  File "/data/DSE/megatron/training.py", line 478, in train
    save_checkpoint(iteration, model, optimizer, lr_scheduler)
  File "/data/DSE/megatron/checkpointing.py", line 116, in save_checkpoint
    save_ds_checkpoint(iteration, model, args)
  File "/data/DSE/megatron/checkpointing.py", line 108, in save_ds_checkpoint
    model.save_checkpoint(args.save, iteration, client_state=sd)
  File "/opt/conda/lib/python3.6/site-packages/deepspeed/runtime/engine.py", line 1418, in save_checkpoint
Traceback (most recent call last):
  File "pretrain_gpt2.py", line 115, in <module>
    self._save_checkpoint(save_dir, tag, client_state=client_state)
    args_defaults={'tokenizer_type': 'GPT2BPETokenizer'})
  File "/opt/conda/lib/python3.6/site-packages/deepspeed/runtime/engine.py", line 1459, in _save_checkpoint
  File "/data/DSE/megatron/training.py", line 101, in pretrain
    train_data_iterator, valid_data_iterator)
  File "/data/DSE/megatron/training.py", line 478, in train
    save_checkpoint(iteration, model, optimizer, lr_scheduler)
  File "/data/DSE/megatron/checkpointing.py", line 116, in save_checkpoint
    save_ds_checkpoint(iteration, model, args)
  File "/data/DSE/megatron/checkpointing.py", line 108, in save_ds_checkpoint
    self._curr_ckpt_path = os.path.join(save_dir, tag)
    model.save_checkpoint(args.save, iteration, client_state=sd)
  File "/opt/conda/lib/python3.6/posixpath.py", line 94, in join
  File "/opt/conda/lib/python3.6/site-packages/deepspeed/runtime/engine.py", line 1418, in save_checkpoint
    genericpath._check_arg_types('join', a, *p)
  File "/opt/conda/lib/python3.6/genericpath.py", line 149, in _check_arg_types
    (funcname, s.__class__.__name__)) from None
TypeError: join() argument must be str or bytes, not 'int'
    self._save_checkpoint(save_dir, tag, client_state=client_state)
  File "/opt/conda/lib/python3.6/site-packages/deepspeed/runtime/engine.py", line 1459, in _save_checkpoint
    self._curr_ckpt_path = os.path.join(save_dir, tag)
  File "/opt/conda/lib/python3.6/posixpath.py", line 94, in join
    genericpath._check_arg_types('join', a, *p)
  File "/opt/conda/lib/python3.6/genericpath.py", line 149, in _check_arg_types
    (funcname, s.__class__.__name__)) from None
TypeError: join() argument must be str or bytes, not 'int'
```

Replacing this line with `model.save_checkpoint(args.save, str(iteration), client_state=sd)` seems to fix the issue.

What is the current status of this pull request? Having this complete would be very helpful for our work.

**Comment by ShadenSmith** [comment]
Thanks, good catch @gcooper-isi . I'm working on examples and will wrap this line up asap :-).

**Pull request merged**